### PR TITLE
Add upload fixed image in badge designer

### PR DESCRIPTION
**Pull request by SegiNyn**
This adds an option to upload images specific to the badge from the badge designer. 

The image can be added from the 'insert elements' dropdown list and is called `Fixed image`

Multiple Fixed images can be added to the badge.

### CODE REVIEW

File: indico/modules/designer/controllers.py
Comment (plourenco) [inline indico/modules/designer/controllers.py:328]: Instead of checking for an input value `for_background`, I would recommend changing a bit the logic here to one of the following:
- We could rename this endpoint to `RHUploadImage` and have another dedicated one for setting the background image `RHSetBackgroundImage`
- Alternatively, we could add a query param `background=True/False` when we intend to set a background. The form action would be: 
```
action="{{ url_for('designer.upload_image', template, background=True) }}"
```

**Review by plourenco — COMMENTED** [review]
Thanks! There were a few ESLint problems, but not worth reviewing those since we can refactor this file in a single commit before merging.

**Comment by plourenco** [comment]
@ThiefMaster since there's currently more than one fixed element, what do you think about having both the `FixedText` and `FixedImage` as placeholders in the designer? 
This could prevent client-side conditionals such as [this](https://github.com/indico/indico/pull/5273/files#diff-b3ff04534fe852620feb9c66bb4e46b192e2e62a8a2d37d067893596755ddf21R118).

File: indico/modules/designer/controllers.py
Comment (plourenco) [inline indico/modules/designer/controllers.py:282]:
```suggestion
    @use_kwargs({'background': fields.Bool(load_default=False)})
    def _process_POST(self):
        data = dict({'background_position': 'stretch', 'items': []}, **request.json['template'])
```

File: indico/modules/designer/controllers.py
Comment (plourenco) [inline indico/modules/designer/controllers.py:329]:
```suggestion
        if background:
            self.template.background_image = image
```

File: indico/modules/designer/controllers.py
Comment (SegiNyn) [reply indico/modules/designer/controllers.py]: Okay, thank you

File: indico/modules/designer/client/js/index.js
Comment (plourenco) [inline indico/modules/designer/client/js/index.js:124]: I think we can simplify this now that we are using placeholders.

```suggestion
        .text(itemTitles[this.type]);
```

**Review by plourenco — APPROVED** [review]

File: indico/modules/designer/client/js/index.js
Comment (SegiNyn) [reply indico/modules/designer/client/js/index.js:124]: I kept this condition because the fixed text can take a user input, making the inputted text different from the item title, also when an image is uploaded for the fixed_image, it should only display the image and not still show the title. So the title should only be used when there is no image.

Comment (plourenco) [reply indico/modules/designer/client/js/index.js:124]: Right, never mind, for some reason I assumed this was referring to the selected text.

**Comment by SegiNyn** [comment]
Hello, please can we merge this PR if it looks good, so we can include it in our next release

**Comment by ThiefMaster** [comment]
Since you need to backport it anyway, I guess it'd be fine for you if we merged it into master and not 3.1.x?

**Comment by SegiNyn** [comment]
Yes, it'll be fine

File: indico/modules/designer/templates/template.html
Comment (ThiefMaster) [inline indico/modules/designer/templates/template.html:121]:
```suggestion
                      action="{{ url_for('designer.upload_image', template, background=true) }}"
```

File: indico/modules/designer/templates/template.html
Comment (ThiefMaster) [inline indico/modules/designer/templates/template.html:277]:
```suggestion
                          action="{{ url_for('designer.upload_image', template, background=false) }}"
```

**Comment by ThiefMaster** [comment]
Unless I'm blind we never check that the `image_id` when saving the badge actually belongs to that particular template... so you could actually "steal" an image from another event by simply sending a different ID...

**Comment by SegiNyn** [comment]
Yes you are correct, there is not check that the image_id belongs to the template. I can add a check for it.

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:302]: we do not use camelCase in python code...

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:195]: Why are we saving a URL in the `data`? This should be generated on the fly...

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:188]: No need for `.get()`, you know the key exists here...

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:187]: When would there be a `fixed_image` item without an `image_id`? If this is not expected to happen, no need to check for it - just let it fail if there's garbage data in the DB.

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:185]: I don't think the `enumerate` is needed, there's no need to use an index. When you iterate over a list of items and mutate an item, it will change the item in the list.

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:184]: Careful! you end up mutating the original template's data here. I think the only reason this may not get saved is that SQLAlchemy doesn't realize it has been mutated...

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:304]: Again, when would there *not* be an `image_id` for a `fixed_image` item?

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:305]: Instead of querying the images on by one, why not build a set with all the images of the template (`{img.id for img in self.template.images}`) and then simply check if every `image_id` is in that set?

File: indico/modules/designer/placeholders.py
Comment (ThiefMaster) [inline indico/modules/designer/placeholders.py:349]: When would this happen? Won't a `fixed_image` *always* have an `image_id`?

File: indico/modules/designer/controllers.py
Comment (SegiNyn) [reply indico/modules/designer/controllers.py:187]: This could happen for example when a user inserts a fixed_image element in the badge designer but they do not upload any image into it. Then the image_id would be null

File: indico/modules/designer/controllers.py
Comment (SegiNyn) [reply indico/modules/designer/controllers.py]: Tbh it's because it was easier to get the url from the item in the javascript. But to change this, would it be okay to include the images url as part of the template_data?

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [reply indico/modules/designer/controllers.py:187]: I think failing validation (so the user has to remove the item or add an image) would be better than storing broken data...

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:184]: you need to make a deepcopy; otherwise you still mutate the individual items..

File: indico/modules/designer/controllers.py
Comment (SegiNyn) [reply indico/modules/designer/controllers.py:187]: Okay

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:304]: Not sure why flake8 didn't complain about this, but those parentheses are certainly not needed...

```suggestion
        template_images = {img.id for img in self.template.images}
```

File: indico/modules/designer/controllers.py
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:308]:
```suggestion
            if image_item['image_id'] not in template_images:
```

File: indico/modules/events/registration/badges.py
Comment (ThiefMaster) [inline indico/modules/events/registration/badges.py:99]: I'd use `elif placeholder.group == 'fixed'` here and else fail with an error (since that'd mean there's a placeholder with an unknown/unhandled group)

File: indico/modules/events/registration/badges.py
Comment (ThiefMaster) [inline indico/modules/events/registration/badges.py:102]: Don't you have the pre-commit.githook enabled? pyupgrade would point out that this should use an f-string:

```suggestion
                    raise ValueError(f'Unknown placeholder group: {placeholder.group}')
```

**Review by ThiefMaster — APPROVED** [review]

File: indico/modules/events/registration/badges.py
Comment (SegiNyn) [reply indico/modules/events/registration/badges.py]: Okay, I will use it for subsequent PRs. Thank you for the reviews.

**Pull request merged**
